File: 1512-design-underground-system/design-underground-system.cpp
class UndergroundSystem {
private:
    unordered_map<string, unordered_map<string, vector<int>>> startToEndTimes;
    unordered_map<int, pair<string, int>> idToStation; 
public:
    UndergroundSystem() {
        startToEndTimes.clear();
        idToStation.clear();
    }
    
    void checkIn(int id, string stationName, int t) {
        idToStation[id] = {stationName, t};
    }
    
    void checkOut(int id, string stationName, int t) {
        pair<string, int> beginToTime = idToStation[id];
        idToStation.erase(id);
        string start = beginToTime.first;
        int prevTime = beginToTime.second;
        startToEndTimes[start][stationName].push_back(t - prevTime);
    }
    
    double getAverageTime(string startStation, string endStation) {
        vector<int> times = startToEndTimes[startStation][endStation];
        int sz = times.size();
        double averageTime = 0;
        int totalTime = 0;
        
        for (int idx = 0; idx < sz; idx++) {
            totalTime += times[idx];
        }

        return (double)totalTime / sz;
    }
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem* obj = new UndergroundSystem();
 * obj->checkIn(id,stationName,t);
 * obj->checkOut(id,stationName,t);
 * double param_3 = obj->getAverageTime(startStation,endStation);
 */

File: 1512-design-underground-system/design-underground-system.py
class UndergroundSystem(object):

    def __init__(self):
        # id mapped with startStation and time
        # startStation mapped with endstation mapped with list of times taken
        self.idStartStationTimeMap = defaultdict()
        self.startEndTimesMap = defaultdict(lambda : defaultdict(list))

    def checkIn(self, id, stationName, t):
        """
        :type id: int
        :type stationName: str
        :type t: int
        :rtype: None
        """
        self.idStartStationTimeMap[id] = (stationName, t)
        

    def checkOut(self, id, stationName, t):
        """
        :type id: int
        :type stationName: str
        :type t: int
        :rtype: None
        """
        startStationAndTime = self.idStartStationTimeMap[id]
        startStation = startStationAndTime[0]
        timeStart = startStationAndTime[1]
        self.startEndTimesMap[startStation][stationName].append(t - timeStart)
        

    def getAverageTime(self, startStation, endStation):
        """
        :type startStation: str
        :type endStation: str
        :rtype: float
        """
        sum , count = 0.0, 0.0

        for time in self.startEndTimesMap[startStation][endStation]:
            sum += time
            count += 1

        return sum / count

# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)